File: tutorial/data/project-walkthrough/pkg-python/analysis/code/03_event_study.py
"""
Analysis Script 03: Event Study

Tests the parallel trends assumption and traces the dynamic effect of the
policy over time by estimating a set of "event-time" dummies relative to
each state's adoption year.

Approach:
  - Compute time_to_treat = year - adoption_year for treated states
  - Bin endpoints: values < -5 are coded as -5; values > 5 are coded as +5
  - Create indicator dummies et_m5, et_m4, ..., et_m1, et_p0, ..., et_p5
  - Omit et_m1 (t = -1) as the reference period so all coefficients are
    relative to the year immediately before adoption
  - Never-treated states (policy_adopted == 0) contribute to identification
    of the year fixed effects as the comparison group

If pre-treatment coefficients (et_m5 through et_m2) are near zero, this
supports the parallel trends assumption underlying the DD design.

Output: analysis/output/figures/event_study.png
"""

import os
from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

try:
    import pyfixest as pf
except ImportError:
    raise ImportError("pyfixest not installed — run: pip install pyfixest")

ROOT = Path(os.environ.get("PROJECT_ROOT", Path(__file__).parent.parent.parent))

# ─── Paths ────────────────────────────────────────────────────────────────────
input_file   = ROOT / "build/output/analysis_panel.csv"
output_plot  = ROOT / "analysis/output/figures/event_study.png"

print("=" * 60)
print("SCRIPT 03: EVENT STUDY")
print("=" * 60)

# ─── Load data ────────────────────────────────────────────────────────────────
print(f"\nReading: {input_file.relative_to(ROOT)}")
df = pd.read_csv(input_file)
print(f"Loaded {len(df):,} state-year observations")

# ─── Log outcome ──────────────────────────────────────────────────────────────
df["log_fatal"] = np.log(df["fatal_crashes"] + 1)

# ─── Time-to-treatment variable ───────────────────────────────────────────────
# For states that never adopted (policy_adopted == 0), adoption_year is NaN.
# Those observations serve as a control group via the year fixed effects and
# do NOT receive event-time dummies.
treated_mask = df["policy_adopted"] == 1

df["time_to_treat"] = np.where(
    treated_mask,
    df["year"] - df["adoption_year"],
    np.nan,
)

# Bin at -5 and +5 so the endpoints absorb all pre/post observations
df["event_time"] = df["time_to_treat"].copy()
df.loc[df["event_time"] < -5, "event_time"] = -5
df.loc[df["event_time"] >  5, "event_time"] =  5

event_times = list(range(-5, 6))  # -5, -4, ..., 0, ..., 5

print(f"\n  Treated states: {df.loc[treated_mask, 'state_fips'].nunique()}")
print(f"  Never-treated states: {df.loc[~treated_mask, 'state_fips'].nunique()}")
print(f"  Event-time range (binned): {min(event_times)} to {max(event_times)}")
print(f"  Reference period: t = -1 (year before adoption)")

# ─── Construct event-time dummies ─────────────────────────────────────────────
# Naming convention: et_mN for negative periods, et_pN for zero/positive.
def et_name(t):
    return f"et_m{abs(t)}" if t < 0 else f"et_p{t}"

for t in event_times:
    df[et_name(t)] = ((df["event_time"] == t) & treated_mask).astype(int)

# Exclude the reference period (t = -1) from the right-hand side
rhs_vars = [et_name(t) for t in event_times if t != -1]
formula = "log_fatal ~ " + " + ".join(rhs_vars) + " | state_fips + year"

print("\n" + "-" * 60)
print("EVENT STUDY REGRESSION")
print(f"  Formula: {formula}")
print("  SE type: cluster-robust (CRV1) by state_fips")
print("-" * 60)

mod = pf.feols(formula, data=df, vcov={"CRV1": "state_fips"})
mod.summary()

# ─── Extract coefficients ─────────────────────────────────────────────────────
# Add the reference period back as a zero row so the plot passes through zero.
rows = []
for t in event_times:
    if t == -1:
        rows.append({"t": t, "coef": 0.0, "se": 0.0})
    else:
        v = et_name(t)
        rows.append({"t": t, "coef": mod.coef()[v], "se": mod.se()[v]})

coef_df = pd.DataFrame(rows).sort_values("t").reset_index(drop=True)
coef_df["ci_lo"] = coef_df["coef"] - 1.96 * coef_df["se"]
coef_df["ci_hi"] = coef_df["coef"] + 1.96 * coef_df["se"]

print("\n" + "-" * 60)
print("EVENT STUDY COEFFICIENTS (t=-1 is reference period, coef=0)")
print("-" * 60)
print(coef_df[["t", "coef", "se", "ci_lo", "ci_hi"]].to_string(index=False))

# ─── Plot ─────────────────────────────────────────────────────────────────────
fig, ax = plt.subplots(figsize=(10, 6))

# Shaded confidence-interval ribbon
ax.fill_between(
    coef_df["t"],
    coef_df["ci_lo"],
    coef_df["ci_hi"],
    alpha=0.20,
    color="steelblue",
    label="95% CI",
)

# Connecting line
ax.plot(coef_df["t"], coef_df["coef"], color="steelblue", linewidth=1.5, zorder=2)

# Point estimates
ax.scatter(coef_df["t"], coef_df["coef"], color="steelblue", s=50, zorder=3,
           label="Point estimate")

# Reference lines
ax.axhline(0, color="black", linewidth=0.9)
ax.axvline(-0.5, color="firebrick", linestyle="--", linewidth=1.5, alpha=0.8,
           label="Policy adoption")

ax.set_xticks(event_times)
ax.set_xlabel("Years Relative to Policy Adoption", fontsize=12)
ax.set_ylabel("Coefficient (log fatal crashes)", fontsize=12)
ax.set_title("Event Study: Dynamic Effect of Policy on Fatal Crashes",
             fontsize=13, pad=14)
ax.legend(frameon=True)
ax.grid(True, alpha=0.25, linestyle="--")

plt.tight_layout()
output_plot.parent.mkdir(parents=True, exist_ok=True)
plt.savefig(output_plot, dpi=300, bbox_inches="tight")
plt.close()

print(f"\nSaved: {output_plot.relative_to(ROOT)}")
print("  Reference period: t = -1 (coef fixed at 0)")
print("  CI ribbon: +/- 1.96 * SE")

File: tutorial/data/project-walkthrough/pkg-r/analysis/code/03_event_study.R
# 03_event_study.R
# Event study (dynamic DiD)
#
# Instead of a single post_treated coefficient, we estimate a separate
# effect for each year relative to policy adoption. This does two things:
#   1. Tests the parallel trends ASSUMPTION: pre-period coefficients should
#      be near zero (no pre-trends).
#   2. Traces out how the policy EFFECT EVOLVES over time.
#
# Time-to-treatment variable:
#   time_to_treat = year - adoption_year  (NA for never-treated states)
# Reference category: t = -1 (one year before adoption)
# Endpoints binned at -5 and +5 to avoid sparse tails.
#
# Outputs:
#   analysis/output/figures/event_study.png

cat("Running 03_event_study.R...\n")

# ------------------------------------------------------------------
# Setup — master.R sets these; edit below if running standalone
if (!exists("root")) {
  root     <- "."
  build    <- file.path(root, "build")
  analysis <- file.path(root, "analysis")
  if (!require("pacman")) install.packages("pacman")
  pacman::p_load(data.table, fixest, ggplot2)
}
# ------------------------------------------------------------------

pacman::p_load(data.table, fixest, ggplot2)

# ---- Read data -------------------------------------------------------
dt <- fread(file.path(root, "build/output/analysis_panel.csv"))
dt[, log_fatal := log(fatal_crashes + 1)]

# ---- Create time_to_treat and ever_treated ---------------------------
# ever_treated = 1 if a state ever adopts, 0 if never-treated.
# never-treated states get time_to_treat = -1000 (a sentinel value that
# fixest's i() function can exclude from the interaction dummies).
dt[, ever_treated := as.integer(!is.na(adoption_year))]

dt[, time_to_treat := fifelse(
  ever_treated == 1,
  as.integer(year - adoption_year),
  -1000L   # sentinel: never-treated, contributes to FE estimation only
)]

# ---- Bin endpoints to avoid sparse tails ----------------------------
# Observations at t <= -5 are grouped into a single "= -5" cell,
# and t >= 5 into a single "= +5" cell. This keeps the plot clean.
dt[time_to_treat != -1000 & time_to_treat < -5, time_to_treat := -5L]
dt[time_to_treat != -1000 & time_to_treat >  5, time_to_treat :=  5L]

# ---- Event study regression ------------------------------------------
# i(time_to_treat, ever_treated, ref = c(-1, -1000)) creates dummies
# for each value of time_to_treat interacted with ever_treated,
# omitting t = -1 (reference) and the never-treated sentinel (-1000).
#
# The never-treated states are included in the regression; they identify
# the year fixed effects but receive a zero on every event-time dummy.
es_model <- feols(
  log_fatal ~ i(time_to_treat, ever_treated, ref = c(-1, -1000)) |
    state_fips + year,
  data = dt,
  vcov = ~state_fips
)

cat("\nEvent Study Coefficients:\n")
print(summary(es_model))

# ---- Extract coefficients for plotting -------------------------------
# iplot() from fixest can do this automatically, but we build the plot
# manually with ggplot2 so we have full control over the aesthetics.
coef_names <- names(coef(es_model))

# Pull the time values out of the coefficient names
# (they look like "time_to_treat::-5:ever_treated", etc.)
coef_dt <- data.table(
  name = coef_names,
  coef = coef(es_model),
  se   = se(es_model)
)
coef_dt <- coef_dt[grepl("time_to_treat", name)]

# Parse the time index from the name string
coef_dt[, time := as.integer(
  gsub(".*time_to_treat::(-?[0-9]+).*", "\\1", name)
)]

# Add the reference period (t = -1, coef = 0 by construction)
coef_dt <- rbind(
  coef_dt[, .(time, coef, se)],
  data.table(time = -1L, coef = 0, se = 0)
)
coef_dt <- coef_dt[order(time)]

# 95% confidence intervals
coef_dt[, ci_lower := coef - 1.96 * se]
coef_dt[, ci_upper := coef + 1.96 * se]

# ---- Plot ------------------------------------------------------------
event_plot <- ggplot(coef_dt, aes(x = time, y = coef)) +
  # Shaded confidence band (ribbon is preferred over error bars)
  geom_ribbon(aes(ymin = ci_lower, ymax = ci_upper),
              fill = "steelblue", alpha = 0.20) +
  # Point estimates connected by a line
  geom_line(color = "steelblue", linewidth = 0.8) +
  geom_point(color = "steelblue", size = 2.5) +
  # Reference lines
  geom_hline(yintercept = 0, linetype = "dashed", color = "black") +
  geom_vline(xintercept = -0.5, linetype = "dashed",
             color = "firebrick", linewidth = 0.5) +
  scale_x_continuous(breaks = -5:5) +
  labs(
    x       = "Years relative to policy adoption",
    y       = "Effect on log fatal crashes",
    caption = "Reference period: t = \u22121. 95% confidence intervals shown.\nEndpoints binned at \u22125 and +5. Standard errors clustered by state."
  ) +
  theme_minimal(base_size = 12) +
  theme(
    axis.title  = element_text(size = 11),
    axis.text   = element_text(size = 10),
    plot.caption = element_text(size = 8, color = "grey40")
  )

# ---- Save figure -----------------------------------------------------
dir.create(file.path(root, "analysis/output/figures"),
           showWarnings = FALSE, recursive = TRUE)

ggsave(
  file.path(root, "analysis/output/figures/event_study.png"),
  plot   = event_plot,
  width  = 9,
  height = 5.5,
  dpi    = 300
)

cat("\n  Saved event_study.png\n")
